sys/stdio_nimble/stdio_nimble.c

```diff
@@ -0,0 +1,399 @@
+/*
+ * Copyright (C) 2019 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys
+ * @{
+ *
+ * @file
+ * @brief       STDIO over NimBLE implementation
+ *
+ *
+ * @author      Hendrik van Essen <hendrik.ve@fu-berlin.de>
+ *
+ * @}
+ */
+
+#include <errno.h>
+#include <stdlib.h>
+
+#include "nimble_riot.h"
+#include "net/bluetil/ad.h"
+
+#include "host/ble_hs.h"
+#include "host/util/util.h"
+#include "host/ble_gatt.h"
+#include "services/gap/ble_svc_gap.h"
+#include "services/gatt/ble_svc_gatt.h"
+
+#include "stdio_uart.h"
+#include "periph/uart.h"
+#include "isrpipe.h"
+
+#include "ringbuffer.h"
+#include "stdio_nimble.h"
+
+/* TODO: Why is 261 the absoulte maximum before it stops working? */
+#define MTU_MAX_NIMBLE 261 //BLE_ATT_MTU_MAX
+
+/*
+ * PDU = LL Header (2 bytes)
+ *     + L2CAP Header (4 bytes)
+ *     + Op Code (1 byte)
+ *     + Attribute Handler (2 bytes)
+ *     + MIC (Optional, 4 bytes)
+ *     = 13 bytes
+ * */
+#define PDU_HEADERS_SIZE 13
+#define MAX_MTU_PAYLOAD (MTU_MAX_NIMBLE - PDU_HEADERS_SIZE)
+
+/* ringbuffer for stdin */
+static ringbuffer_t _rb_stdin;
+static char _rb_stdin_mem[CONFIG_STDIO_NIMBLE_STDIN_BUFSIZE];
+
+/* ringbuffer for stdout */
+static ringbuffer_t _rb_stdout;
+static char _rb_stdout_mem[CONFIG_STDIO_NIMBLE_STDOUT_BUFSIZE];
+
+static mutex_t _rb_stdout_mutex = MUTEX_INIT;
+static int _current_max_mtu_payload = MAX_MTU_PAYLOAD;
+
+/* intermediate buffer to transfer data between ringbuffer and nimble functions,
+ * which are all based on os_mbuf implementation */
+static char _stdin_read_buf[MAX_MTU_PAYLOAD];
+static char _stdout_write_buf[MAX_MTU_PAYLOAD];
+
+/* information about bluetooth connection */
+static uint16_t _conn_handle;
+static uint16_t _val_handle_stdout;
+static bool _send_stdio = false;
+static bool _sending_in_progress = false;
+static mutex_t _send_mutex = MUTEX_INIT;
```

Comment:
Can we get rid of this mutex and use `tsrb`?

Reply:
We can get rid of `_send_mutex` but we still need `_tsrb_stdout_mutex` even with `tsrb` as `tsrb_free` and `tsrb_add` should be used as atomic operation here, because multiple calls of `stdio_write` in parallel are possible.